include/eosio/vm/vector.hpp

```diff
@@ -1,97 +1,110 @@
 #pragma once
 
 #include <eosio/vm/exceptions.hpp>
+#include <eosio/vm/allocator.hpp>
 
 #include <utility>
 #include <string>
+#include <vector>
 
 namespace eosio { namespace vm {
-   template <typename T, typename Allocator> 
-   class managed_vector {
-      public:
-         constexpr managed_vector(Allocator& allocator, size_t size=0) :
-            _size(size),
-            _allocator(&allocator),
-            _data(allocator.template alloc<T>( _size )) {
-         }
-
-         constexpr managed_vector(const managed_vector& mv) = delete;
-         constexpr managed_vector(managed_vector&& mv) {
-            _size  = mv._size;
-            _allocator = mv._allocator;
-            _data  = mv._data;
-         }
-
-         constexpr managed_vector& operator=(managed_vector&& mv) {
-            _size  = mv._size;
-            _allocator = mv._allocator;
-            _data  = mv._data;
-            return *this;
-         }
-
-         constexpr inline void set_owner( Allocator& alloc ) { _allocator = &alloc; }
-         constexpr inline void resize( size_t size ) {
-            if (size > _size) {
-               T* ptr = _allocator->template alloc<T>( size );
-               if (_size == 0)
-                 _data = ptr;
-            } else {
-               _allocator->template reclaim<T>( _data + size, _size - size );
+   namespace detail {
+      template <typename T, typename Allocator> 
+      class vector {
+         public:
+            constexpr vector(Allocator& allocator, size_t size=0) :
+               _size(size),
+               _allocator(&allocator),
+               _data(allocator.template alloc<T>( _size )) {
+            }
+
+            constexpr vector(const vector& mv) = delete;
+            constexpr vector(vector&& mv) = default;
+            constexpr vector& operator=(vector&& mv) = default;
+
+            constexpr inline void resize( size_t size ) {
+               if (size > _size) {
+                  T* ptr = _allocator->template alloc<T>( size );
+                  if (_size == 0)
+                    _data = ptr;
+               } else {
+                  _allocator->template reclaim<T>( _data + size, _size - size );
+               }
+               _size = size;
+            }
+
+            constexpr inline void emplace_back( T&& val ) {
+               EOS_VM_ASSERT( _index < _size, wasm_vector_oob_exception, "vector write out of bounds" );
+               _data[_index++] = std::forward<T>(val);
```

Comment:
std::forward is a bit pointless, given that this implementation requires T to have a trivial default constructor/copy constructor/destructor.

Reply:
I'm aware, I didn't push up a lot of different tries because of the excess overhead.